### PR TITLE
Preserve GMPO loss when use_liger_kernel is enabled

### LINKED ISSUE #7270: [Bug] GMPOTrainer with use_liger_kernel=True trains on the GRPO objective
### Reproduction

With `use_liger_kernel=True`, `GMPOTrainer` silently trains on the GRPO objective. The reproduction below compares a full optimizer update with the GMPO and GRPO objectives.

Related: [Issue #6808](https://github.com/huggingface/trl/issues/6808) already reports the dispatch problem; [PR #7163](https://github.com/huggingface/trl/pull/7163) proposes rejecting the configuration at initialization, and [PR #6817](https://github.com/huggingface/trl/pull/6817) was an earlier attempt.

`GMPOTrainer` overrides `_compute_loss` but inherits `compute_loss` from `GRPOTrainer`, which dispatches before `_compute_loss` is reached:

```python
def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
    if return_outputs:
        raise ValueError("The GRPOTrainer does not support returning outputs")
    if self.use_liger_kernel:
        unwrapped_model = self.accelerator.unwrap_model(model)
        return self._forward_redirection(model, unwrapped_model, self.compute_liger_loss, unwrapped_model, inputs)
    return self._compute_loss(model, inputs)
```

So with `use_liger_kernel=True`, GMPO trains on the fused GRPO loss.

To reproduce: `GMPOTrainer.train` with a two-layer LLaMA (hidden size 64, vocabulary 128), a fixed initial checkpoint and a fixed rollout generated by a separate behavior policy, one SGD update (lr `0.05`, momentum `0.9`), once with `use_liger_kernel=False` and once with `True`.

The Liger-enabled GMPO run is identical to a plain GRPO run in loss, full gradient and final state. Compared with the GMPO objective, its gradient differs by `0.1996558210` relative L2. Overriding `compute_loss` in `GMPOTrainer` to call its own `_compute_loss` brings that to `1.93049e-7`; GRPO runs with and without Liger serve as controls for ordinary kernel differences. Reproduced on a second host.

Expected: an execution flag should either keep the GMPO objective or reject the configuration, not silently switch to GRPO. Rejecting the flag as in #7163 is a reasonable API choice. Routing to the ordinary GMPO loss, as tested here, keeps the objective but not the fused kernel's memory savings. This is about TRL's integration, not the GMPO paper code, and the one-update test says nothing about long-run reward.

### System Info

```text
TRL: 32fc9d71333c0390bed5b74051526aab0361441e
PyTorch: 2.8.0+cu128
Transformers: 5.17.0
liger-kernel: 0.8.3
Triton: 3.4.0
GPU: NVIDIA RTX A6000
```

Source: [gmpo_trainer.py](https://github.com/huggingface/trl/blob/32fc9d71333c0390bed5b74051526aab0361441e/trl/experimental/gmpo/gmpo_trainer.py), [grpo_trainer.py](https://github.com/huggingface/trl/blob/32fc9d71333c0390bed5b74051526aab0361441e/trl/trainer/grpo_trainer.py).

### Checklist

- [ ] I have checked that my issue isn't already filed (the related issue #6808 reports this problem; this report adds a full-update reproduction).
- [x] I have included my system information
- [ ] Any code provided is minimal, complete, and reproducible (a code excerpt and reproduction protocol are included; a standalone executable reproducer is not attached).
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete (no exception; this is a silent objective switch)

### PR DESCRIPTION
**Pull request by gss10282025**
## What does this PR do?

Fixes #7270. `GMPOTrainer.compute_loss` now always dispatches to its geometric-mean objective, including when `use_liger_kernel=True`.

The inherited GRPO dispatcher currently sends that flag through the fused GRPO loss and bypasses GMPO's `_compute_loss`. This override preserves the return-output check and profiling wrapper, and routes both flag values through GMPO.

Related: #6808 and #7163. #7163 proposes rejecting the flag at initialization. This draft is an alternative: keep the ordinary GMPO loss when the flag is set. It does not add a fused GMPO kernel or its memory savings, and the inherited constructor's Liger availability/configuration checks still apply. If rejecting the flag is the preferred API, #7163 is the simpler choice.

## Validation

`python -m pytest tests/experimental/test_gmpo_loss_dispatch.py -q`: **3 passed** on CPU. The regression exercises the actual dispatcher and GMPO loss for both flag values, compares loss and gradients with an independent geometric-mean reference, and checks `return_outputs=True` rejection. It stubs model log-probability production and does not initialize a Liger kernel.

Ruff lint and formatting passed. CUDA/Liger end-to-end training has not been run on this branch.

## Before submitting

- [x] Read the contributor guideline.
- [x] Linked the existing discussion and related PR above; no maintainer approval is claimed.
- [x] Updated the GMPO documentation.
- [x] Added regression tests.

## AI writing disclosure

- [ ] No AI usage: entirely human-written.
- [ ] AI-assisted and reviewed by a human.
- [x] AI-generated: implementation and tests prepared by a coding agent.

## Who can review?

Draft pending review of the fallback behavior and native training validation.

---

> [!NOTE]
> **Medium Risk**
> Changes RL training loss dispatch for GMPO when Liger is enabled; incorrect behavior would silently train with the wrong objective, though new tests guard the dispatch path.
> 
> **Overview**
> **`GMPOTrainer` now overrides `compute_loss`** so training always uses the geometric-mean `_compute_loss`, even when `use_liger_kernel=True`. Previously, inheriting `GRPOTrainer`'s dispatcher could route through the fused GRPO objective and skip GMPO's loss.
> 
> The override keeps the profiling wrapper and rejects `return_outputs=True` with a clear error. Docs note that Liger does not provide a fused GMPO kernel—only the standard GMPO loss runs, without fused memory savings.
> 
> Regression tests assert both flag values match an independent geometric-mean reference (loss and gradients) and that fused GRPO paths are never invoked.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit eaef2248ee47313ff0bd1384d212649458ef7e2b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).
